File: app/src/main/java/com/fallgamlet/dnestrcinema/utils/ViewUtils.java
package com.fallgamlet.dnestrcinema.utils;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.support.annotation.NonNull;
import android.support.design.widget.Snackbar;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.widget.Toast;

import com.fallgamlet.dnestrcinema.R;


public class ViewUtils {

    public static void hideKeyboard(Activity activity, View v) {
        if (activity == null) {
            return;
        }

        if (v == null) {
            v = activity.getWindow().getCurrentFocus();
        }

        if (v == null) {
            return;
        }

        hideKeyboard((Context) activity, v);
    }

    public static void hideKeyboard(Context context, View view) {
        if (context == null || view == null) {
            return;
        }

        InputMethodManager imm = (InputMethodManager) context.getSystemService(Context.INPUT_METHOD_SERVICE);
        imm.hideSoftInputFromWindow(view.getWindowToken(), 0);
    }

    public static void showKeyboard(Activity activity, View v) {
        if (activity == null) {
            return;
        }

        if (v == null) {
            v = activity.getWindow().getCurrentFocus();
        }

        if (v == null) {
            return;
        }

        showKeyboard(activity, v);
    }

    public static void showKeyboard(Context context, View view) {
        if (context == null || view == null) {
            return;
        }

        InputMethodManager imm = (InputMethodManager) context.getSystemService(Context.INPUT_METHOD_SERVICE);
        imm.showSoftInput(view, 0);
    }

    public static void showToast(Context context, String text) {
        Toast.makeText(context, text, Toast.LENGTH_SHORT)
             .show();
    }

    public static void showSnackbar(@NonNull View view, String text) {
        Snackbar.make(view, text, Snackbar.LENGTH_SHORT)
                .setAction("Action", null)
                .show();
    }

    public static void shareApp(Context context) {
        StringBuilder builder = new StringBuilder();
        builder.append(context.getString(R.string.app_name));
        builder.append(" \n ");
        builder.append("https://play.google.com/store/apps/details?id=");
        builder.append(context.getString(R.string.app_id));

        String message = builder.toString();

        share(context, message);
    }

    public static void share(Context context, String message) {
        try {
            Intent sendIntent = new Intent();
            sendIntent.setAction(Intent.ACTION_SEND);
            sendIntent.putExtra(Intent.EXTRA_TEXT, message);
            sendIntent.setType("text/plain");
            context.startActivity(Intent.createChooser(sendIntent, "Поделиться"));
        } catch (Exception e) {
            LogUtils.log("Share", "Fail then sharing", e);
        }
    }


    public static void setVisible(@NonNull View view, boolean visible) {
        setVisible(view, visible, View.GONE);
    }

    public static void setVisible(@NonNull View view, boolean visible, int invisibleState) {
        view.setVisibility(visible? View.VISIBLE: invisibleState);
    }

}

File: app/src/main/java/com/fallgamlet/dnestrcinema/utils/ViewUtils.kt
package com.fallgamlet.dnestrcinema.utils

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.content.res.ColorStateList
import android.os.IBinder
import android.view.View
import android.view.inputmethod.InputMethodManager
import android.widget.TextView
import android.widget.Toast
import androidx.annotation.StringRes
import androidx.core.content.ContextCompat
import androidx.core.view.ViewCompat
import com.fallgamlet.dnestrcinema.R
import com.google.android.material.snackbar.Snackbar


object ViewUtils {

    @JvmOverloads
    fun hideKeyboard(activity: Activity?, v: View? = null) {
        val view = v
            ?: activity?.window?.currentFocus
            ?: activity?.window?.decorView

        hideKeyboard(activity as Context?, view)
    }

    fun hideKeyboard(context: Context?, view: View?) {
        context ?: return
        val v = view ?: View(context)

        try {
            v.postDelayed({
                val binder = v.windowToken
                hideKeyboard(context, binder)
            }, 1)
        } catch (ignored: Exception) { }
    }

    private fun hideKeyboard(context: Context, binder: IBinder?) {
        binder ?: return
        val imm = context.getSystemService(Context.INPUT_METHOD_SERVICE) as? InputMethodManager
        imm?.hideSoftInputFromWindow(binder, InputMethodManager.HIDE_NOT_ALWAYS)
        imm?.hideSoftInputFromInputMethod(binder, InputMethodManager.HIDE_NOT_ALWAYS)
    }


    fun showKeyboard(activity: Activity?, view: View?) {
        val v = view
            ?: activity?.window?.currentFocus
            ?: return

        showKeyboard(activity as? Context, v)
    }

    fun showKeyboard(context: Context?, view: View?) {
        view ?: return
        context ?: return

        val imm = context.getSystemService(Context.INPUT_METHOD_SERVICE) as? InputMethodManager
        imm?.showSoftInput(view, 0)
    }

    fun showToast(context: Context?, text: CharSequence) {
        context ?: return
        Toast.makeText(context, text, Toast.LENGTH_SHORT)
            .show()
    }

    fun showSnackbar(view: View, @StringRes resId: Int) {
        val text = view.context.getString(resId)
        showSnackbar(view, text)
    }

    fun showSnackbar(view: View, text: CharSequence, duration: Int = Snackbar.LENGTH_SHORT) {
        makeSnackbar(view, text, duration)
            .show()
    }

    fun makeSnackbar(view: View, text: CharSequence, duration: Int = Snackbar.LENGTH_SHORT): Snackbar {
        val context = view.context
        val snackBar = Snackbar.make(view, text, duration)
        val snackBarView = snackBar.view

        val bgColor = ContextCompat.getColor(context, R.color.colorTextPrimary)
        ViewCompat.setBackgroundTintList(snackBarView, ColorStateList.valueOf(bgColor))

        snackBarView.findViewById<TextView>(com.google.android.material.R.id.snackbar_text)
            ?.setTextColor(ContextCompat.getColor(context, R.color.colorTextLight))

        return snackBar
    }

    fun shareApp(context: Context) {
        val message = context.getString(R.string.app_name) +
                "\n https://play.google.com/store/apps/details?id=${context.packageName}"

        share(context, message)
    }

    fun share(context: Context, message: String) {
        try {
            val title = context.getString(R.string.share)
            val sendIntent = Intent()
            sendIntent.action = Intent.ACTION_SEND
            sendIntent.putExtra(Intent.EXTRA_TEXT, message)
            sendIntent.type = "text/plain"
            context.startActivity(Intent.createChooser(sendIntent, title))
        } catch (e: Exception) {
            LogUtils.log("Share", "Fail then sharing", e)
        }
    }

}
